tools/generate-force-kernels.py

```diff
@@ -0,0 +1,383 @@
+        assert self.norm  == 2.0 * s
+
+    def __call__(self, w):
+        return 2 * 1.0 / self.norm * numpy.sum(numpy.sin((i+1) * w) * c for i, c in enumerate(self.coeffs))
+
+def decic(k, v):
+    """ Deconvolves the CIC window, once"""
+    f = 1
+    for dir in range(3):
+        cellsize = (v.BoxSize[dir] / v.Nmesh[dir])
+        wd = k[dir] * cellsize
+        tmp = numpy.sinc(wd * 0.5 / numpy.pi)
+        f = f * 1 / tmp ** 2
+    return v * f
+
+# low noise differentiator
+def gradient(dir, dfkernel):
+    def kernel(k, v, df=dfkernel, dir=dir):
+        cellsize = (v.BoxSize[dir] / v.Nmesh[dir])
+        w = k[dir] * cellsize
+        a = dfkernel(w) / cellsize
+        return v * (1j * a)
+    return kernel
+
+def longrange(r_split):
+    if r_split != 0:
+        def kernel(k, v):
+            kk = sum(ki ** 2 for ki in k)
+            return v * numpy.exp(-kk * r_split**2)
+    else:
+        def kernel(k, v):
+            return v
+    return kernel
+
+def laplace(k, v):
+    kk = sum(ki ** 2 for ki in k)
+    mask = (kk == 0).nonzero()
+    kk[mask] = 1
+    b = v / kk
+    b[mask] = 0
+    return b
+
+def force_pm(pm, x, y, dfkernel=NDiff('lnld4_5'), split=1.25, compensate_cic=True):
+    """ split is in unit of mesh cells, 1.25 is Gadget Asmth.
+        lnld4_5 is gadget differentiation.
+    """
+    rho = pm.paint(y)
+    x = numpy.array(x)
+    p = rho.r2c()\
+        .apply(laplace, out=Ellipsis)
+
+    if compensate_cic:
+        p = p.apply(decic, out=Ellipsis) \
+             .apply(decic, out=Ellipsis) # twice, once for paint, once of readout
+
+    if split:
+        p = p.apply(longrange(split * pm.BoxSize[0] / pm.Nmesh[0]), out=Ellipsis)
+
+    r = []
+    for d in range(3):
+        f = p.apply(gradient(d, dfkernel=dfkernel))
+        f = f.c2r(out=Ellipsis)
+
+        r.append(f.readout(x))
+
+    p = p.c2r(out=Ellipsis)
+    p = p.readout(x)
+
+    # Add 4 pi t omatch the direct sum.
+    return numpy.array(r).T * (4 * numpy.pi), p * (4 * numpy.pi)
+
+# brute force with summation of images,
+
+from itertools import product
+
+def gravity_spline(dist, a):
+    # copied from Gadget / check for typos? But we really only care the very outside
+    r = numpy.einsum('...j,...j->...', dist, dist) ** 0.5
+    fac = 1 / r ** 3
+    pot = - 1 / r
+
+    u = r / a
+
+    inner = u < 0.5
+    mid   = (u >= 0.5) & (u < 1.0)
+    fac[inner] = (a ** -3 * ( 32. / 3 + u * u * (32.0 * u - 38.4)))[inner]
+    fac[mid] = (a ** -3 * ( 64. / 3 - 48.0 * u
+                        + 38.4 * u * u
+                        - 32.0 / 3 * u * u * u
+                        - 0.2/3 / (u * u * u)))[mid]
+
+    pot[inner] = (a**-1 * (
+            -2.8 + u * u * (16./3 + u * u * (6.4 * u - 0.6))
+        ))[inner]
+    pot[mid]  = (a**-1 * (
+            -3.2 + 0.2 / 3 / u + u * u * (32. / 3 + u * (-16.0 + u *(9.6 - 6.4 / 3 * u)))
+        ))[mid]
+
+    return dist * fac[..., None], -pot
+
+def gravity_plummer(dist, a):
```

Comment:
I understand where gravity_plummer comes from, but it's not totally obvious from the name. gravity_spline makes sense if you happen to be familiar enough with the Gadget source to know recognise the softening kernel is, but could become confusing otherwise (especially if we change the softening kernel). Maybe you could add some comments?